### PR TITLE
core, tests: remove visa for etcs trains

### LINKED ISSUE #11461: core: remove VISA for ETCS trains
Confidential issue: https://github.com/osrd-project/osrd-confidential/issues/951

The goal is to disable the VISA for ETCS equiped trains

Acceptance criteria:
- when a user asks for a stop for a train, on a ETCS equiped track, the train has to follow the ETCS bracking curves (see [related issue](https://github.com/osrd-project/osrd-confidential/issues/1004))
- when there's a SvL (Supervised Location), there is no VISA, but a release speed associated to the SvL, shown in the speed-diagram

### PR DESCRIPTION
**Pull request by louisgreiner**
Close https://github.com/OpenRailAssociation/osrd/issues/11461

Respects the ACs:
- VISA disabled for stops
- VISA disabled for last buffer at end of the path (removing short slip)

Updated `test_etcs_schedule_result_slowdowns` integration test.

New integration test created, `test_etcs_schedule_result_slowdowns_with_stop`, see Space Speed Chart (first stop on closed signal, last stop on open signal followed by buffer-stop):
- before this PR
with visa - with stop
- after this PR
no visa - with stop

### CODE REVIEW

**Comment by louisgreiner** [comment]
Question : should we keep the `etcs_bal` infra ? or should I revert these commits:
- https://github.com/OpenRailAssociation/osrd/pull/11694/commits/e36415ca757be3511163b24ae19e180446c8871c
- https://github.com/OpenRailAssociation/osrd/pull/11694/commits/d34c1853867407ccd4545fbd9697647993508a18

**Comment by louisgreiner** [comment]
@bougue-pe @eckter 
Question: should we keep the `etcs_bal` infra? Or should I revert the associated commits? I am not sure of the relevance of storing the whole `.railjson` in the code.

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt [hunk 1]
```diff
@@ -49,12 +51,32 @@ fun makeSafetySpeedRanges(
                     it.receptionSignal == RJSTrainStop.RJSReceptionSignal.SHORT_SLIP_STOP
                 )
             }
+            .filter {
+                // Stops that are in a range of ETCS_LEVEL2 signaling system are handled later
```
Comment (bougue-pe) [inline core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt]:
```suggestion
                // ETCS signaling system already handles Safety Approach Speed via braking curves
```

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt [hunk 2]
```diff
@@ -49,12 +51,32 @@ fun makeSafetySpeedRanges(
...
+                !isStopInSignalingSystemRange(
+                    it.offset,
+                    signalingRanges,
+                    ETCS_LEVEL2.id,
+                )
+            }
             .toMutableList()
-    makeEndOfPathStop(rawInfra, routes, signalOffsets)?.let { stopsWithSafetySpeed.add(it) }
+
+    // Last buffer stop is handled separately when in a ETCS_LEVEL2 signaling system range
```
Comment (bougue-pe) [inline core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt]:
```suggestion
    // Last buffer-stop is handled via ETCS braking curves if the last stop is handled by ETCS signaling system
```

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt [hunk 3]
```diff
@@ -73,6 +95,24 @@ fun makeSafetySpeedRanges(
...
+    for (range in signalingRanges) {
+        if (
+            range.value == signalingSystem &&
+                range.lower <= stopOffset.distance &&
+                range.upper >= stopOffset.distance
+        ) {
+            return true
+        }
+    }
+    return false
```
Comment (bougue-pe) [inline core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt]: Actually, there is the `get()` on DistanceRangeMap:
```suggestion
    return (signalingRanges.get(stopOffset) == signalingSystem)
```

**Review by bougue-pe — COMMENTED** [review]
Didn't review tests yet, but the 'prod' part LGTM with just details.

And I ❤️ the mixed infra!

File: core/src/test/kotlin/fr/sncf/osrd/standalone_sim/StandaloneSimulationTest.kt [hunk 4]
```diff
@@ -355,6 +358,55 @@ class StandaloneSimulationTest {
...
+                    42.seconds,
+                    STOP,
+                ),
+                // Stop that will be skipped from safety speed range (in ETCS_LEVEL2 range),
+                // associated to the 10150m signal
+                SimulationScheduleItem(
+                    Offset(10_000.meters),
+                    42.seconds,
+                    42.seconds,
+                    STOP,
```
Comment (bougue-pe) [inline core/src/test/kotlin/fr/sncf/osrd/standalone_sim/StandaloneSimulationTest.kt]: nit: we could make it a short-slip stop to be more exhaustive on tests
```suggestion
                    SHORT_SLIP_STOP,
```

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt (hunk 1 above)
Comment (Erashin) [reply core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt]: We could even add it's through the release speed.

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt [hunk 5]
```diff
@@ -49,12 +51,32 @@ fun makeSafetySpeedRanges(
                     it.receptionSignal == RJSTrainStop.RJSReceptionSignal.SHORT_SLIP_STOP
                 )
             }
+            .filter {
```
Comment (Erashin) [inline core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt]: You could do this filter before the map (either with a second filter or just by adding it to the previous filter).

**Review by Erashin — COMMENTED** [review]
Good luck with the rebase, sorry again 😅

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt [hunk 6]
```diff
@@ -49,12 +51,32 @@ fun makeSafetySpeedRanges(
...
+    if (
+        schedule.isNotEmpty() &&
+            !isStopInSignalingSystemRange(
+                schedule.last().pathOffset,
+                signalingRanges,
+                ETCS_LEVEL2.id,
+            )
+    ) {
+        makeEndOfPathStop(rawInfra, routes, signalOffsets)?.let { stopsWithSafetySpeed.add(it) }
+    }
```
Comment (eckter) [inline core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt]: This block could be simplified if we first build a list with all `SafetySpeedStop` (including the last one), and *then* filter out the ones in ETCS ranges

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt [hunk 7]
```diff
@@ -49,12 +51,32 @@ fun makeSafetySpeedRanges(
...
 
     val res = distanceRangeMapOf<Speed>()
     for (stop in stopsWithSafetySpeed) {
         // Currently, safety speed is applied to the next signal no matter the sight distance
+        // TODO: should we compute the minimum when adding a speed limit?
```
Comment (eckter) [inline core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt:81]: What kind of minimum would you add?

The distances are truncated at the end of the method, and the speeds are compared with the rest of the speeds for the MRSP

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt (hunk 6 above)
Comment (bougue-pe) [reply core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt]: Thought about that, but this one is the buffer-stop, and probably not part of the travelled-path (unless stopping right at the buffer-stop).
And the filtering on this "VISA-place" is made considering the signaling range of the last stop, not of the buffer-stop.

So the last stop can trigger 2 VISA (if BAL signaling): on next signal and on buffer-stop.

We could still handle this during filtering (if a stop is after travelled-path, check the end of the travelled-path), but this is not so straightforward. So current code or suggested change (tweaked) are fine by me.

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt (hunk 1 above)
Comment (louisgreiner) [reply core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt]: Done

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt (hunk 2 above)
Comment (louisgreiner) [reply core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt]: Done thanks

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt (hunk 3 above)
Comment (louisgreiner) [reply core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt]: Much better :)

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt (hunk 1 above)
Comment (bougue-pe) [reply core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt]: I disagree about the release speed: to me this is abusive language that it "manages VISA in ETCS", as the VISA is a max-speed constraint, and the release-speed is actually a release of the max-speed constraint (so the opposite).

What makes speed safe for the protected location in ETCS is that we are following EBD-based curves that include a release speed because we consider it's safe enough to let the driver responsible under that speed. And without release-speed, we would even be safer.
What make speed safe for the BAL case is that we are following CONST braking to be at 30kph 200m before protected location.

So I'd stay to "braking curves" (that includes release-speed, which is considered safe enough, but release-speed is not made for safety but for operations).

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt (hunk 7 above)
Comment (bougue-pe) [reply core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt:81]: The example I can think of:
StopA 1000m SHORT_SLIP_STOP
StopB 1050m STOP
Signal 1100m

With the current behavior I guess it's bugged:
- StopA adds a 30kph range on 900-1000m + 10kph range on 1000-1100m
- Then StopB overwrites with a 30kph range 900-1100m

Comment (louisgreiner) [reply core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt:81]: ~~It was a thought we had with @bougue-pe that I forgot to remove. It indeed seems that we don't need to compute the minimum here, since the MRSP handles it~~

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt (hunk 5 above)
Comment (louisgreiner) [reply core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt:54]: I tried this implementation, proposed below:
```kt
    val allStopsWithSafetySpeed = schedule
        .filter { it.receptionSignal.isStopOnClosedSignal }
        .map {
            SafetySpeedStop(
                it.pathOffset,
                it.receptionSignal == RJSTrainStop.RJSReceptionSignal.SHORT_SLIP_STOP
            )
        }.toMutableList()

    makeEndOfPathStop(rawInfra, routes, signalOffsets)?.let { allStopsWithSafetySpeed.add(it) }

    // ETCS signaling system already handles Safety Approach Speed via braking curves (Release Speed)
    val stopsWithSafetySpeed = allStopsWithSafetySpeed.filter {
        !isStopInSignalingSystemRange(it.offset, signalingRanges, ETCS_LEVEL2.id)
    }.toMutableList()
```

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt (hunk 7 above)
Comment (eckter) [reply core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt:81]: Oh good point @bougue-pe, I missed this.

It shouldn't be that tricky to fix, either we run two loops (one for all the 30km/h limits and then the 10km/h). Or we can call `updateMapIntersection` to check the minimum.

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt (hunk 6 above)
Comment (eckter) [reply core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt]: Oh! Very good point. Yeah my suggestion won't work, sorry.

File: core/src/test/kotlin/fr/sncf/osrd/standalone_sim/StandaloneSimulationTest.kt (hunk 4 above)
Comment (louisgreiner) [reply core/src/test/kotlin/fr/sncf/osrd/standalone_sim/StandaloneSimulationTest.kt]: Done

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt (hunk 5 above)
Comment (louisgreiner) [reply core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt]: Indeed, merged into the first filter

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt (hunk 1 above)
Comment (Erashin) [reply core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt]: Works for me :)

**Review by Erashin — APPROVED** [review]
GJ. I'll let the other reviewers handle their comments/approvals :)

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt (hunk 7 above)
Comment (eckter) [reply core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt:81]: I've opened a new issue: https://github.com/OpenRailAssociation/osrd/issues/11768

It's a bug (from before this PR), but it's extremely unlikely to cause any issue.

I'm closing this thread.

**Review by eckter — APPROVED** [review]

File: core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt (hunk 7 above)
Comment (bougue-pe) [reply core/src/main/kotlin/fr/sncf/osrd/standalone_sim/SafetySpeed.kt:81]: I think `updateMap` is a better candidate, considering that there are empty parts on the initial map.
And I have a preference for this indeed (a bit more generic).

Anyway, it's not a must for this PR to be merged IMO (weird corner case, not changing the current behavior).
But it would be nice in a follow-up PR.

File: core/src/test/kotlin/fr/sncf/osrd/standalone_sim/StandaloneSimulationTest.kt [hunk 8]
```diff
@@ -355,6 +358,55 @@ class StandaloneSimulationTest {
...
+                SimulationScheduleItem(
+                    Offset(10_300.meters),
+                    42.seconds,
+                    42.seconds,
+                    STOP,
+                ),
+            )
+
+        val signalingRanges = buildSignalingRanges(infra, blocks.toIdxList(), chunkPath)
+        signalingRanges.put(9_500.meters, 10_400.meters, ETCS_LEVEL2.id)
```
Comment (bougue-pe) [inline core/src/test/kotlin/fr/sncf/osrd/standalone_sim/StandaloneSimulationTest.kt:397]: (May be in a follow-up PR) I'd like to have some dedicated corner-case tests:
* ETCS from 9500 to 10350: should keep the same behavior (no VISA)
* BAL all the way, but ETCS on 10100 to 10400: should have VISA ending on signal (10150) and not on buffer-stop (10400)
* BAL all the way, and ETCS from 10350 to 10400: should have a VISA "everywhere" (10150 + 10400)
* An initial stop at 100m under ETCS would be nice too to ensure "happy-path" behavior.

File: tests/infra-scripts/etcs_bal_infra.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+This is derived from small_infra.
+"""
+
+import sys
+from pathlib import Path
+
+from small_infra_creator import create_small_infra
+
+signaling_systems = {
+    "WS": "ETCS_LEVEL2",
```
Comment (bougue-pe) [inline tests/infra-scripts/etcs_bal_infra.py]: So happy with this, but I see 2 points:
* This is unused currently, so this should be removed from this PR and added in a follow-up PR:
  * probably replacing etcs_infra with WS, SWS and maybe MWS in BAL
  * then testing a bit on SES->SWS for BAL braking + VISA)
* The 2 commits (.py + .json) should be squashed to have CI passing on every commit

**Review by bougue-pe — APPROVED** [review]
Thanks !

✅ locally tested on the mixed-infra, GEV is sooo nice with VISA/no-VISA depending on the sections (and const/gamma braking curves too) 😍 

To sum-up, to me this is mergeable, but I'd rather:
* remove the mixed-infra from this PR
* open a follow-up PR (add mixed-infra, add integration tests on BAL with mixed-infra, add some more unit-tests, maybe #11768)

File: tests/infra-scripts/etcs_bal_infra.py (hunk 9 above)
Comment (louisgreiner) [reply tests/infra-scripts/etcs_bal_infra.py]: Tracked in https://github.com/OpenRailAssociation/osrd/issues/11783, created PR https://github.com/OpenRailAssociation/osrd/pull/11776 to continue this

File: core/src/test/kotlin/fr/sncf/osrd/standalone_sim/StandaloneSimulationTest.kt (hunk 8 above)
Comment (louisgreiner) [reply core/src/test/kotlin/fr/sncf/osrd/standalone_sim/StandaloneSimulationTest.kt:397]: Created this issue to do it in a follow-up PR : https://github.com/OpenRailAssociation/osrd/issues/11777
